packages/beacon-node/src/chain/seenCache/seenBlockInput.ts

```diff
@@ -0,0 +1,253 @@
+      //     sampledColumns: this.custodyConfig.sampledColumns,
+      //     source: {
+      //       source,
+      //       seenTimestampSec,
+      //       peerIdStr
+      //     }
+      //   })
+      // }
+      else {
+        blockInput = BlockInputBlobs.createFromBlock({
+          block: block as SignedBeaconBlock<ForkBlobsDA>,
+          blockRootHex,
+          daOutOfRange,
+          forkName,
+          source: {
+            source,
+            seenTimestampSec,
+            peerIdStr,
+          },
+        });
+      }
+      this.blockInputs.set(blockInput.blockRootHex, blockInput);
+    }
+
+    if (!blockInput.hasBlock()) {
+      blockInput.addBlock({block, blockRootHex, source: {source, seenTimestampSec, peerIdStr}});
+    } else {
+      this.logger?.debug("Attempt to cache block but is already cached on BlockInput", blockInput.getLogMeta());
+      this.metrics?.seenCache.blockInput.duplicateBlockCount.inc();
+    }
+
+    return blockInput;
+  }
+
+  getByBlob(
+    {blobSidecar, source, seenTimestampSec, peerIdStr}: SourceMeta & {blobSidecar: deneb.BlobSidecar},
+    opts: GetByBlobOptions = {}
+  ): BlockInputBlobs {
+    const blockRoot = this.config
+      .getForkTypes(blobSidecar.signedBlockHeader.message.slot)
+      .BeaconBlockHeader.hashTreeRoot(blobSidecar.signedBlockHeader.message);
+    const blockRootHex = toRootHex(blockRoot);
+
+    // TODO(peerDAS): Why is it necessary to static cast this here. All conditional paths result in a valid value so should be defined correctly below
+    let blockInput = this.blockInputs.get(blockRootHex) as IBlockInput;
+    let created = false;
+    if (!blockInput) {
+      created = true;
+      const {forkName, daOutOfRange} = this.buildCommonProps(blobSidecar.signedBlockHeader.message.slot);
+      blockInput = BlockInputBlobs.createFromBlob({
+        blobSidecar,
+        blockRootHex,
+        daOutOfRange,
+        forkName,
+        source,
+        seenTimestampSec,
+        peerIdStr,
+      });
+      this.metrics?.seenCache.blockInput.createdByBlob.inc();
+      this.blockInputs.set(blockRootHex, blockInput);
+    }
+
+    if (!isBlockInputBlobs(blockInput)) {
+      throw new SeenBlockInputCacheError(
+        {
+          code: SeenBlockInputCacheErrorCode.WRONG_BLOCK_INPUT_TYPE,
+          cachedType: blockInput.type,
+          requestedType: DAType.Blobs,
+          ...blockInput.getLogMeta(),
+        },
+        `BlockInputType mismatch adding blobIndex=${blobSidecar.index}`
+      );
+    }
+
+    if (!blockInput.hasBlob(blobSidecar.index)) {
+      blockInput.addBlob({blobSidecar, blockRootHex, source, seenTimestampSec, peerIdStr});
+    } else if (!created) {
+      this.logger?.debug(
+        `Attempt to cache blob index #${blobSidecar.index} but is already cached on BlockInput`,
+        blockInput.getLogMeta()
+      );
+      this.metrics?.seenCache.blockInput.duplicateBlobCount.inc();
```

Comment:
would be great to add "source" to this metric so that we know where does the duplicate come from